[FIX] l10n_at: do not delete all tags on update

wtaferner opened the pull request:
**Description of the issue/feature this PR addresses:**
As given by a test migration the template will somehow be applied
on related account.account which will remove all manually or from other
modules applied tags.

**Current behavior before PR:**
For unknown reasons (as it is closed source code of Odoo SA), the migration is deleting manually applied tags and also tags owned by another module which is not available during migration

**Desired behavior after PR is merged:**
Migration does not remove any tags which are not directly related to a template change.

Info: @wt-io-it



--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/79602).

wtaferner commented:
@beledouxdenis @jco-odoo 
May I ask you to review and merge this one. Denis and I were kinda identifying this as a possible cause for data loss during migration. Not sure if this is the case as I am kinda blind and have no insight to the code causing this issue, but at least linking only is a less intrusive approach to data. I also think that the template to real chart update is at a non-mature level and not officially supported yet. Obviously during migration something is tried like that.

jco-odoo commented:
@wtaferner I don't see immediately how this can impact the existing account.account or even the migration.  But in itself, this change does not seem problematic.

wtaferner commented:
@jco-odoo this is why I asked @beledouxdenis before but I guess those who have done the migration scripts might know whether this is needed and wanted or the migration scripts need to be refined which I still think is needed but out of my scope and possibilities.

beledouxdenis commented:
Think out of scope of the upgrade. Even in standard, in stable, each time you do a a `-u l10n_at`, or even a -u `account` or -u `base`, your tags from `l10n_at_account_tag_hierarchy` gets deleted when this file is loaded, because of the `[(6,)]` tuple commands, then added back when `l10n_at_account_tag_hierarchy` is loaded. Meaning the tags assigned on the account templates are being removed and added back over and over again at each `-u` , which is not very optimistic.

Note that with this, I am afraid we could have the same tags added multiple times on the account templates, each time the file is loaded it adds again the same tags. Either that, either it raises a unique constraint when loading this file.

I could alter the upgrade scripts so it removes tags coming from standard modules only, but still I think this fact it removes the tags from the account templates and add them back all over again at each `-u` in stable/standard is really not efficient.

william-andre commented:
> For unknown reasons (as it is closed source code of Odoo SA)

Because the upgrade process is updating the module. That means loading the xml file again, and therefore running the x2many commands.

______

@beledouxdenis @wtaferner @jco-odoo 
In this case, there is no issue in using a `4` command because you can't duplicate the data. But in other models, creating sub lines with a  `0` command, it would duplicate the lines at each upgrade. That is why a `5` command is done first on the creation of `account.tax.template` for the repartitions lines for instance. (it is only for templates so it is mostly okay)

beledouxdenis commented:
> 
> 
> > For unknown reasons (as it is closed source code of Odoo SA)
> 
> Because the upgrade process is updating the module. That means loading the xml file again, and therefore running the x2many commands.


To be complete, we are speaking about this upgrade script:
https://github.com/odoo/upgrade/blob/master/migrations/account/0.0.0/end-account-tags.py

which tries to smartly assign the tags on the accounts according to the changes occurring on the templates during the `-u` of the `l10n` modules. Basically, it computes the difference on the tags assigned on the account templates before and after the `-u`, and then add and removes the tags on the accounts themselves matching the account template codes.

In this case, because of the `[(6, )]` tuple commands, any tags added after this module, for instance in the case of Wolfgang the tags added with his module `l10n_at_account_tag_hierarchy`, are being removed from the templates, and therefore this upgrade scripts which tries to do things smartly removes these tags from the accounts themselves also.

I see two solutions:
- either in standard we assign the tags using `[(4, )]` commands instead of `[(6, )]` commands. This has the advantage to avoid having this back and forth to always remove and add back the tags during a `-u` in stable
- either in upgrade, we remove tags coming from standard modules only. This has the advantage we won't need to alter each and every `l10n` data files doing `[(6, )]` commands instead of `[(4, )]`

william-andre commented:
> To be complete, we are speaking about this upgrade script:
> https://github.com/odoo/upgrade/blob/master/migrations/account/0.0.0/end-account-tags.py

Wait. Why are we looking at the templates during the upgrade? That seems wrong: once a chart of accounts has been loaded it should never look at the templates again. If changes need to be done, it should be done manually (maybe through a server action or something, but not automatically during the upgrade).

wtaferner commented:
> Wait. Why are we looking at the templates during the upgrade?

This was actually my first thought, but who knows what is going on in the Odoo SA Community 😉

beledouxdenis commented:
I just explained William through internal communication, but basically the goal is to handle those cases of changes:
https://github.com/odoo/odoo/commit/c75fe95dfca366bf6c3b6a51908677039488aa4f
https://github.com/odoo/odoo/commit/02f2691bffaa419fb61c41c66e04320a76220eb5

where the tags and the accounting reports are completely refactored, and there was no upgrade script coming with them to remap the tags.

As nothing was done to re-assign the tags on the existing accounts, the accounting reports were be empty or partially wrong.

Alternatively, yes, an upgrade script could be written along the refactoring commit to map the former tags to the new tags on each account.

But, at the end, this potential script doing the map would be longer than this automatic script doing things smartly by analyzing the changes in the templates and applying them in the accounts. At the end it does automatically what you would have to do manually if you had to write the mapping upgrade script yourself. You would note the differences applied in the XML files of the account templates and apply the same difference on the `account.account` instances, using the account codes for matching.

I will write the patch so it doesn't remove the tags coming from custom modules. Though I am still convinced than spending time on resetting the tags list all over again at each `-u` is really not ideal.

wtaferner commented:
@beledouxdenis 
So it was you and your aim to improve the world...very much appreciated 🙂 
I would suggest to go into a refactoring that this `account.account.template` -> `account.account` update can be done by the accounting user and therefore actively monitored and maybe even mapped as I think it is even wanted sometimes in the same version to get a sync with the latest templates.

Not sure if the fully automatic approach would really resolve the issue you guys have during migration, but yeah, if the approach is water-proof I do appreciate every automation to cleanup data as Murphy and users are one 🤣

william-andre commented:
> I would suggest to go into a refactoring that this `account.account.template` -> `account.account` and also the `account.tax.template` -> `account.tax` update can be done by the accounting user and therefore actively monitored and maybe even mapped as I think it is even wanted sometimes in the same version to get a sync with the latest templates.

We are working on it.

wtaferner commented:
@jco-odoo @william-andre @beledouxdenis 
I guess based on the information given it should be merged as it would help Denis to handle the dirty use cases and in case we need to update the template as a tag is completely wrong we can explicitly remove tags instead of a general removal of all tags before adding those only given in the module...

@beledouxdenis if you are not faster than this PR, let me know when the migration scripts are fixed that I can do a new test migration to proof it is actually working as expected.

beledouxdenis commented:
The PR in odoo/upgrade is created. A little review from my peers and we are good :)

beledouxdenis commented:
The PR in upgrade is merged, you can replay your upgrade

wtaferner commented:
@beledouxdenis 
Ok, without this being merged or do I need this change that it works?

wtaferner commented:
@jco-odoo @william-andre 
Alright, the upgrade issue is more or less resolved. Now I added a view things (operating, financing tags) which got deleted because it was not in the template which I improved additionally to the change from the 6 to 4 command. I also removed the useless empty tax_ids with the 6 command.

Ready to merge from my side...fire at will 😉

william-andre commented:
The changes done to convert the commands 6,SET to 4,LINK are useless, make the diff unreadable, and are obfuscating the history.
The rest of the changes don't apply to the commit message.

The commands 6,SET should be used to manage the case when we want to remove a tag from an tax template. If we remove the 4,LINK command, it wouldn't do anything. We try to keep this way of doing everywhere for the templates to have common ground in the way the l10n modules work.
This is master data and shouldn't be changed manually by the user. If another module is changing that, just update that module afterwards. The original issue of the custom tags being removed is already fixed in the upgrade script.

wtaferner commented:
Ok, so I do a separate PR for the other stuff and the changes 6 to 4 are not needed (I agree).
